src/backend/distributed/planner/multi_logical_planner.c

```diff
@@ -1510,12 +1510,12 @@ TargetListOnPartitionColumn(Query *query, List *targetEntryList)
 
 	/*
 	 * We could still behave as if the target list is on partition column if
-	 * all range table entries are reference tables and all target expressions
-	 * are column references to the given query level.
+	 * all range table entries are reference tables or intermediate results,
+	 * and all target expressions are column references to the given query level.
 	 */
 	if (!targetListOnPartitionColumn)
 	{
-		if (RangeTableListContainsOnlyReferenceTables(query->rtable) &&
+		if (RangeTableListContainsOnlyRecurringTuples(query->rtable) &&
```

Comment:
The `read_intermediate_result` function is typically wrapped in a subquery. Would we still encounter it in the `rtable`?

Reply:
`TargetListOnPartitionColumn ()` is called for every `Query` node in the subquery. So, The following should also work:
```SQL
BEGIN;
 
SELECT broadcast_intermediate_result('squares', 'SELECT s, s*s FROM generate_series(1,200) s');

-- single appereance of read_intermediate_result
SELECT 
  DISTINCT user_id 
FROM 
  users_table 
JOIN (
	SELECT *,random() FROM (SELECT 
	  max(res.val) as mx 
	FROM 
  		read_intermediate_result('squares', 'binary') AS res (val int, val_square int) 
	GROUP BY res.val_square) foo)
squares
 ON (mx = user_id)
ORDER BY 1
LIMIT 5;
```

Is that what you mean? If so, I can add a regression test for it.

Reply:
I mean that the way in which `read_intermediate_result` is used is:
```
SELECT 
  DISTINCT user_id 
FROM 
  users_table 
JOIN (
SELECT *,random() FROM (SELECT 
  max(res.val) as mx 
FROM 
  (SELECT val, val_square FROM read_intermediate_result('squares', 'binary') AS res (val int, val_square int)) res
GROUP BY res.val_square) foo)
squares
 ON (mx = user_id)
ORDER BY 1
LIMIT 5;
ERROR:  cannot push down this subquery
DETAIL:  Group by list without partition column is currently unsupported
```

Reply:
Hmm, could we actually end up with such query? As far as I see & test, we keep recursively planning until it we get rid of `GROUP BY`s?
```SQL
SELECT DISTINCT user_id
FROM users_table
JOIN
  (SELECT *,
          random()
   FROM
     (SELECT max(res.val) AS mx
      FROM
        (SELECT users_table.value_2 AS val,
                users_table.value_2*19 AS val_square
         FROM users_table
         GROUP BY 1,
                  2) res
      GROUP BY res.val_square) foo) squares ON (mx = user_id)
ORDER BY 1 ;
DEBUG:  replacing subquery SELECT value_2 AS val, (value_2 * 19) AS val_square FROM public.users_table GROUP BY value_2, (value_2 * 19) --> SELECT val, val_square FROM read_records_file('base/pgsql_job_cache/10_0_8877_1_0.data'::text, 'binary'::text) read_records_file(val integer, val_square integer)
DEBUG:  replacing subquery SELECT max(val) AS mx FROM (SELECT read_records_file.val, read_records_file.val_square FROM read_records_file('base/pgsql_job_cache/10_0_8877_1_0.data'::text, 'binary'::text) read_records_file(val integer, val_square integer)) res GROUP BY val_square --> SELECT mx FROM read_records_file('base/pgsql_job_cache/10_0_8877_1_1.data'::text, 'binary'::text) read_records_file(mx integer)
```

Reply:
Man, recursive planning is pretty cool :), though I guess we'd avoid the recursion if we'd know how to handle the GROUP BY here?

Reply:
Or, why do we need this PR then 😄  I guess, it's still OK to have this for completeness in any way.

> though I guess we'd avoid the recursion if we'd know how to handle the GROUP BY here?

Yes, it probably makes sense. So, the rule would be `if (RangeTableListContainsOnlyRecurringTuples) -> stop recursion`

Reply:
> Yes, it probably makes sense. 

Ah, but that'd complicate the code here.

Reply:
Not sure what you mean by `stop recursion`. This code gets called by `DeferErrorIfCannotPushdownSubquery`, which gets called by `ShouldRecursivelyPlanSubquery`. If this code decides that there is no problem with the GROUP BY on the subquery then we do not recursively plan that subquery.

Reply:
> though I guess we'd avoid the recursion if we'd know how to handle the GROUP BY here?

Yes, we could. But, this function is called for every `Query`in the input query. To understand the above case, we'd also need to recursively go into the `RTE_SUBQUERY`s in, which seems to complicate things. That's why I'd probably keep it simple for now. What do you think?

Reply:
well, recursive planning is much more expensive than descending into the query again here, maybe add a TODO.

Reply:
Actually, I'm not considering the cost here. 

What I'm more concerned is that subquery pushdown error checks (i.e., `DeferErrorIfCannotPushdownSubquery()`) are done for every query, and I thought we've avoided any recursion within the checks. 

But, now I remembered that we had to recursively do the reference table checks to prevent pushing down cases like:

```SQL
SELECT * FROM
(SELECT user_id FROM users_ref_table) as u1 LEFT JOIN
(SELECT user_id FROM users_table) as u2 ON (u1.user_id = u2.user_id);
```

I think the query we discuss has very similar characteristics and seems like it'd be OK to call a function name `HasOnlyRecurringTuples()`, which is similar to `HasRecurringTuples` but returns true only when all the `RTEs` recurring tuples (i.e., Recursive version of `RangeTableListContainsOnlyRecurringTuples ()`).

Does that make sense to you as well? If yes, I can add a `TODO` comment here.

Reply:
> Does that make sense to you as well?

Yes, ok. We're kind of missing such a function.

Reply:
I'll add that to this PR

Reply:
I added the function and the test you've mentioned above. See `HasNonRecurringTuples()` function, it is slightly different than what I expected at first. But, seems simpler to implement.